packages/react/src/xcss-prop/index.ts

```diff
@@ -0,0 +1,143 @@
+ */
+export type CompiledStyles<TObject> = {
+  [Q in keyof TObject]: TObject[Q] extends Record<string, unknown>
+    ? CompiledStyles<TObject[Q]>
+    : CompiledPropertyDeclarationReference;
+};
+
+/**
+ * Please think twice before using this type, you're better off declaring explicitly
+ * what your API should be, for example only defining `color`.
+ *
+ * Use in conjunction with {@link XCSSProp} to allow all properties to be given to
+ * your component.
+ */
+export type XCSSAllProperties = keyof CSSProperties;
+
+/**
+ * Please think twice before using this type, you're better off declaring explicitly
+ * what your API should be, for example not allowing any pseudos at all using the
+ * `never` type.
+ *
+ * Use in conjunction with {@link XCSSProp} to allow all pseudos to be given to
+ * your component.
+ */
+export type XCSSAllPseudos = CSSPseudos;
+
+/**
+ * ## xcss prop
+ *
+ * Declare styles your component takes with all other styles marked as violations
+ * by the TypeScript compiler. There are two primary use cases for xcss prop:
+ *
+ * - safe style overrides
+ * - inverting style declarations
+ *
+ * Interverting style declarations is interesting for platform teams as
+ * it means products only pay for styles they use as they're now the ones who declare
+ * the styles!
+ *
+ * The {@link XCSSProp} type has generics which must be defined — of which should be what you
+ * explicitly want to maintain as API. Use {@link XCSSAllProperties} and {@link XCSSAllPseudos}
+ * to enable all properties and pseudos.
+ *
+ * @example
+ * ```
+ * interface MyComponentProps {
+ *   // Color is accepted, all other properties / pseudos are considered violations.
+ *   xcss?: XCSSProp<'color', never>;
+ *
+ *   // Only backgrond color and hover pseudo is accepted.
+ *   xcss?: XCSSProp<'backgroundColor', '&:hover'>;
+ *
+ *   // All properties are accepted, all pseudos are considered violations.
+ *   xcss?: XCSSProp<XCSSAllProperties, never>;
+ *
+ *   // All properties are accepted, only the hover pseudo is accepted.
+ *   xcss?: XCSSProp<XCSSAllProperties, '&:hover'>;
+ * }
+ *
+ * function MyComponent({ xcss }: MyComponentProps) {
+ *   return <div css={{ color: 'var(--ds-text-danger)' }} className={xcss} />
+ * }
+ * ```
+ *
+ * The xcss prop works with static inline objects and the [cssMap](https://compiledcssinjs.com/docs/api-cssmap) API.
+ *
+ * @example
+ * ```
+ * // Declared as an inline object
+ * <Component xcss={{ color: 'var(--ds-text)' }} />
+ *
+ * // Declared with the cssMap API
+ * const styles = cssMap({ text: { color: 'var(--ds-text)' } });
+ * <Component xcss={styles.text} />
+ * ```
+ *
+ * To concatenate and conditonally apply styles use the {@link cssMap} {@link cx} functions.
+ */
+export type XCSSProp<
+  TAllowedProperties extends keyof CSSProperties,
+  TAllowedPseudos extends CSSPseudos
+> =
+  | (XCSSItem<TAllowedProperties> & XCSSPseudos<TAllowedProperties, TAllowedPseudos> & XCSSAtRules)
```

Comment:
We've got a `XCSSAtRules` used which is `{ […]?: never }`, should this be a defaulted generic if we want this by default, eg.:
```suggestion
  TAllowedPseudos extends CSSPseudos,
  TAllowedAtRules extends XCSSAtRules = XCSSAtRules
> =
  | (XCSSItem<TAllowedProperties> & XCSSPseudos<TAllowedProperties, TAllowedPseudos> & TAllowedAtRules)
```

Not sure what the plan to get AtRules working is all.  Maybe this is more of a TODO piece?

Reply:
Yeah it's a TODO in a follow up. For now we completely block it. It's because `css` `cssMap` have different APIs for the same thing — need to agree on a path forward.

Reply:
(If we didn't define it, it would mean anyone can pass anything! Which we want to avoid)